### ISSUE TITLE
Blocking until window updated content

### ISSUE THREAD

**Issue by panmari** [fyne-io/fyne#207]
I have a multi threaded program that renders an image. During rendering, I'd like to show a preview window of the image being rendered, updating every X seconds.

Thanks to another issue filed on github, I found that I can use `w.Canvas().Refresh(pw.w.Content())` to trigger a refresh of the window's content. I'm calling this continuously in a go routine:

```go
go func() {
	for {
		time.Sleep(2 * time.Second)
		w.Canvas().Refresh(pw.w.Content())
	}
}()
```

Now the problem: This only works if I stop my rendering threads (e.g. with `time.Sleep(100 * time.Milisecond`) every once in a while. Otherwise my rendering threads use all CPU time and the refresh is never rendered in the window.

Is there a blocking call for `Canvas().Refresh(..)`, so I can manually stop my rendering threads until the content is refreshed?

Or is there an entirely different approach that could work here?

**Comment by andydotxyz** [comment]
This is a really interesting question and the right solution is not immediately obvious.
I would imagine that rendering a single frame, then calling refresh, then rendering another would be the right approach - with the current code you could be rendering lots of frames that are never displayed, or calling refresh when the frame has not changed.

Just a note that canvas.Refresh(obj) is also available, and you should pass the lowest scope possible (i.e. the image itself not the whole window content) for best performance.

Reading the message I think that there is unexpected thread behaviour as the renderer should not be stopped because of many background threads...

**Comment by jadr2ddude** [comment]
For your multithreaded renderer, how do you transfer the image out? If you could use channels, you could use a raster and have the callback read from this channel, while your renderer uses a select

**Comment by jadr2ddude** [comment]
Also, if you are locking up the go runtime, you may have to just call https://golang.org/pkg/runtime/#Gosched so that other stuff can run

**Comment by andydotxyz** [comment]
I guess I should have asked also which type of image/raster you are using and, as @jadr2ddude  asks, how the new image data is being set?

**Comment by panmari** [comment]
Thanks for your input! I tried to address all of it below:

1. [Scope of refresh] Thanks for the tip! I changed my code from `w.Canvas().Refresh(pw.w.Content())` to `canvas.Refresh(pw.img)`
2. [Transfer image out] Shared memory! All render threads (and now, the fyne window/canvas), share a common instance of a struct that implements image.Image. Race conditions don't occur because each render thread operates on a different part of the image.
3. [Multiple frames] There's only one frame that takes a while to render (ray-tracing is slow...)
4. [Using Gosched] Good idea! How can I guarantee that the window-update-goroutine is continued, not another rendering-goroutine? Is there a way to give goroutines a priority?
5. [Type of fyne image] I'm using `img := canvas.NewImageFromImage({shared struct from 2})`

**Comment by andydotxyz** [comment]
2. Shared memory gets tricky - if you run with the go race checker you might see some warnings.
3. In this case don't refresh on a timer - you could do so each time a frame is complete?
5. Assuming you want pixel perfect render I would recommend a [Raster](https://godoc.org/fyne.io/fyne/canvas#NewRaster) instead - that asks for a frame of image data each refresh.

Given the time taken to render a frame I would imagine that saving the pixels (as in your shared memory) and then reading this from the raster generator func on render may be the most efficient. If your buffer is a image.Image then you could use NewRasterFromImage to save some boilerplate code.

If you make those changes do you still find that the UI does not refresh correctly?

**Comment by andydotxyz** [comment]
Have you had any success with this?

**Comment by panmari** [comment]
Sorry, I didn't have the time to follow up on this. I'll try to find some time next week.

**Comment by andydotxyz** [comment]
No pressure, I just wanted to check that we don't have issues going stale :)

**Comment by marcoValkyriweb** [comment]
> 
> 
> Also, if you are locking up the go runtime, you may have to just call https://golang.org/pkg/runtime/#Gosched so that other stuff can run

Thanks, solved my issue of how to get multi line content updated, using a channel to get the info.

**Comment by andydotxyz** [comment]
> getting Fyne to run was also a <2 minute operation

Glad to hear it :)

**Comment by panmari** [comment]
I made it work thanks to 
1. Your pointers on how to to minimize work of a refresh.
2. Optimize threading in my application.

To go into more details for 2: Initially, I spawned a new go routine for every chunk of work
```go
for _, t := range task {
  go func() {
    // Do lots of work here
  }
}
```
I changed this to use a worker pool instead
```go
	taskChan := make(chan task, 128)

	for w := 0; w < runtime.NumCPU(); w++ {
		go worker(taskChan)
	}

	for _, t := range tasks {
		taskChan <- t
	}
```

This helped with the thread contention and made the UI update more frequently. Thanks again for your help!
